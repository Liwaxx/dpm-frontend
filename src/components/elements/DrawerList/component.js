import React from 'react';
import Badge from '@material-ui/core/Badge';
import Collapse from '@material-ui/core/Collapse';
import Divider from '@material-ui/core/Divider';
import ExpandLess from '@material-ui/icons/ExpandLess';
import ExpandMore from '@material-ui/icons/ExpandMore';
import Icon from '@material-ui/core/Icon';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import ListSubheader from '@material-ui/core/ListSubheader';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import DrawerList from './index';
import IconLogtan from '../Icon';

export default class Component extends React.Component {
  constructor() {
    super();

    this.state = {
      open: false
    };
    this._handleClick = this._handleClick.bind(this);
  }

  _handleClick() {
    return () => {
      this.setState({ open: !this.state.open });
    };
  }

  _renderNavItem(isChild = false) {
    const { isMiniDrawer, item } = this.props;
    const { pathname } = location;
    let onClick = () => { };
    let expand = null;
    let title = '';
    let color = pathname.toLowerCase().match(item.url) && item.url ? 'primary' : 'disabled';
    let icon = null;

    if (isChild) {
      onClick = this._handleClick;
      expand = this.state.open ? <ExpandLess /> : <ExpandMore />;
    }
    
    if (isMiniDrawer) {
      title = item.name;
    }

    if (item.isDefault && pathname == '/') {
      color = 'primary';
    }

    if (item.icon) {
      icon = <Icon color={color} title={title}>{item.icon}</Icon>;
    } else if (item.customIcon) {
      const isActive = color == 'primary';

      icon = <IconLogtan isActive={isActive} name={item.customIcon} title={title} />;
    }

    return (
      <ListItem className="drawer-list__expand" button onClick={onClick()}>
        <div>
          {(item.badge) ? <Badge color={item.badge.color ? item.badge.color : 'default'}
            badgeContent={item.badge.content ? item.badge.content : ''}>{icon}
          </Badge> : icon
          }
        </div>
        {(!isMiniDrawer) && <ListItemText inset primary={item.name} />}
        {expand}
      </ListItem>
    );
  }

  _renderTitle() {
    const { item } = this.props;

    return (<ListSubheader>{item.title}</ListSubheader>);
  }

  _renderDivider() {
    return (<Divider />);
  }

  _renderChild() {
    const { isMiniDrawer, item } = this.props;
    const isChild = true;

    return (
      <div>
        {this._renderNavItem(isChild)}
        <Collapse component="li" in={this.state.open} timeout="auto" unmountOnExit>
          <List style={{ paddingLeft: '10px' }}>
            {
              item.children.map((childItem, idx) => (
                <DrawerList item={childItem} isMiniDrawer={isMiniDrawer} key={idx} />
              ))
            }
          </List>
        </Collapse>
      </div>
    );
  }

  _renderNavLink() {
    const { classes, item } = this.props;

    return (
      <Link to={item.url} className={classes.link}>
        {this._renderNavItem()}
      </Link>
    );
  }

  render() {
    const { item } = this.props;

    if (item.title) {
      return this._renderTitle();
    } else if (item.divider) {
      return this._renderDivider();
    } else if (item.children) {
      return this._renderChild();
    } else {
      return this._renderNavLink();
    }
  }
}

Component.propTypes = {
  classes: PropTypes.object,
  isMiniDrawer: PropTypes.bool,
  item: PropTypes.object,
};
