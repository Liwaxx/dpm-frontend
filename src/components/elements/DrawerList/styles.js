const styles = {
  link: {
    textDecoration: 'none'
  },
  listItem: {
    fontSize: '12px',
  }
};

export default styles;
