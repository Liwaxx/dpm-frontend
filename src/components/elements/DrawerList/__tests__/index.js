import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import DrawerList from '../index';
import Component from '../component';

jest.mock('react-router-dom');

describe('DrawerList', () => {
  it('renders item', () => {
    const item = { name: 'test', icon: '', url: '' };
    const tree = renderer
      .create(<DrawerList item={item} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders mini drawer', () => {
    const item = { name: 'test', icon: '', url: '' };
    const tree = renderer
      .create(<DrawerList isMiniDrawer item={item} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders item with badge', () => {
    const item = { name: 'test', icon: '', url: '', badge: { content: '' } };
    const tree = renderer
      .create(<DrawerList item={item} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders item with badge color', () => {
    const item = { name: 'test', icon: '', url: '', badge: { color: 'test', content: 'test' } };
    const tree = renderer
      .create(<DrawerList item={item} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders title item', () => {
    const item = { title: 'test' };
    const tree = renderer
      .create(<DrawerList item={item} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders divider item', () => {
    const item = { divider: true };
    const tree = renderer
      .create(<DrawerList item={item} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('renders child item', () => {
    const item = {
      children: [
        { name: 'test', icon: '', url: '' },
      ]
    };
    const tree = renderer
      .create(<DrawerList item={item} />)
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it('change state when child is expanded', () => {
    const item = {
      children: [
        { name: 'test', icon: '', url: '' },
      ]
    };
    const wrapper = shallow(<Component item={item} />);
    wrapper.find('.drawer-list__expand').first().simulate('click');
    expect(wrapper.state('open')).toEqual(true);
  });
});