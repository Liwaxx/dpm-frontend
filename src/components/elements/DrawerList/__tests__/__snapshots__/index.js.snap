// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerList renders child item 1`] = `
<div>
  <div
    className="MuiButtonBase-root-23 MuiListItem-root-13 MuiListItem-default-16 MuiListItem-gutters-20 MuiListItem-button-21 drawer-list__expand"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    role="button"
    tabIndex="0"
  >
    <div />
    <div
      className="MuiListItemText-root-26 MuiListItemText-inset-27"
    />
    <svg
      aria-hidden="true"
      className="MuiSvgIcon-root-32"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
      />
      <path
        d="M0 0h24v24H0z"
        fill="none"
      />
    </svg>
    <span
      className="MuiTouchRipple-root-39"
    />
  </div>
</div>
`;

exports[`DrawerList renders divider item 1`] = `
<hr
  className="MuiDivider-root-9"
/>
`;

exports[`DrawerList renders item 1`] = `null`;

exports[`DrawerList renders item with badge 1`] = `null`;

exports[`DrawerList renders item with badge color 1`] = `null`;

exports[`DrawerList renders mini drawer 1`] = `null`;

exports[`DrawerList renders title item 1`] = `
<li
  className="MuiListSubheader-root-3 MuiListSubheader-sticky-8 MuiListSubheader-gutters-6"
>
  test
</li>
`;
